rfcs/0128-redeployable-clients.md

```diff
@@ -0,0 +1,393 @@
+
+The clients that do not dynamically generate code still need to retain
+information from the API references and schemas. Those clients should be
+shipped together with a set of references and schemas, but provide the facility
+for these references and schemas to be refetched from a taskcluster deployment
+(specified by a taskcluster root URL).
+
+## Changes to taskcluster client features and configuration
+
+Constructors for clients should require the rootURL to be explicitly provided,
+i.e. no default. This is in order to ensure that the choice of deployed cluster
+to connect to is an active one. The client may provide a utility methods to fetch
+TASKCLUSTER_ROOT_URL from the environment, but it should not be assumed that if
+this environment variable exists, that it should be used. Rather, code that calls
+the taskcluster client should have to explicitly declare they wish to do so, for
+example:
+
+```
+queue := queue.NewFromEnvVars()
+```
+
+This is important because not all code will want to configure settings based on
+environment variables, for example workers that take their configuration from
+configuration files.
+
+## Changes to adding taskcluster client as a service dependency
+
+When upgrading to the new client, care will need to be taken to ensure that
+taskcluster root URL is passed into the client. In addition, the
+references/schemas that the client reads dynamically should also be a
+dependency of the project, which may be either frozen references/schemas,
+fetched from a language package, or fetched dynamically during build/CI from a
+taskcluster deployment.
+
+## Changes to projects that depend on a taskcluster client
+
+Those that depend on a client with generated code based on the content of the
+API references and schemas, should either run the code generation tool inside
+their project on an ad hoc basis, and check in the generated code, or should
+run it as part of their build/ci procedure.
+
+All code that uses any of the new clients will need to explicitly pass in a
+root URL in a constructor, or explicitly call a method to fetch
+`TASKCLUSTER_ROOT_URL` from the environment.
+
+## Changes to `taskcluster-proxy` and its configuration
+
+The taskcluster proxy should require an additional property to start up, which
+is the root url of the deployment it should talk to. This should be a mandatory
+property, and no default should be supplied, so the user is forced to make an
+active choice about which environment they wish the proxy to connect to.
+
+## Changes to workers
+
+* Workers that are started by a provisioner should fetch `TASKCLUSTER_ROOT_URL`
+  from provisioner on start up
+* Workers should require a new config property for taskcluster root URL in
+  their config, and refuse to run if they do not have it and can't fetch it
+  from provisioner configuration (no default allowed)
+* Workers should pass `TASKCLUSTER_ROOT_URL` environment variable to all task
+  processes they create.
```

Comment:
We had talked about whether to provide a TASKCLUSTER_ROOT_URL that points to the proxy (something like `http://localhost`) or to require tasks that want to use the proxy to hard-code some proxy value that we commit to supporting on all workers.  This section doesn't really make the plan clear.

Reply:
Addressed below.